esmvalcore/preprocessor/_volume.py

```diff
@@ -125,33 +124,46 @@ def calculate_volume(cube: Cube) -> da.core.Array:
     """
     # Load depth field and figure out which dim is which
     depth = cube.coord(axis='z')
-    z_dim = cube.coord_dims(depth)[0]
+    z_dim = cube.coord_dims(depth)
+    # Assert z has length > 0
+    if not z_dim:
+        raise ValueError("Cannot compute volume with length 0 Z-axis")
+
+    # Guess bounds if missing
+    if not depth.has_bounds():
+        depth.guess_bounds()
+    if depth.core_bounds().shape[-1] != 2:
+        raise ValueError(
+            f"Z axis bounds shape found {depth.core_bounds().shape}. "
+            "Bounds should be 2 in the last dimension to compute the "
+            "thickness.")
 
     # Calculate Z-direction thickness
     thickness = depth.core_bounds()[..., 1] - depth.core_bounds()[..., 0]
 
-    # Try to calculate grid cell area
-    try:
-        area = da.array(compute_area_weights(cube))
-    except CoordinateMultiDimError:
-        logger.error(
-            "Supplementary variables are needed to calculate grid cell "
-            "areas for irregular grid of cube %s",
-            cube.summary(shorten=True),
-        )
-        raise
+    # Get or calculate the horizontal areas of the cube
+    has_cell_measure = bool(cube.cell_measures('cell_area'))
+    _try_adding_calculated_cell_area(cube)
+    area = cube.cell_measure('cell_area')
+    area_dim = cube.cell_measure_dims(area)
 
-    # Try to calculate grid cell volume as area * thickness
-    if thickness.ndim == 1 and z_dim == 1:
-        grid_volume = area * thickness[None, :, None, None]
-    elif thickness.ndim == 4 and z_dim == 1:
-        grid_volume = area * thickness[:, :]
-    else:
-        raise ValueError(
-            f"Supplementary variables are needed to calculate grid cell "
-            f"volumes for cubes with {thickness.ndim:d}D depth coordinate, "
-            f"got cube {cube.summary(shorten=True)}"
-        )
+    # Make sure input cube has not been modified
+    if not has_cell_measure:
+        cube.remove_cell_measure('cell_area')
+
+    # Get slices to make area and thickness match with the same number
+    # of dimensions as the cube
+    z_slice = tuple(
+        slice(None) if i in z_dim else None
+        for i in range(cube.ndim)
+    )
+    area_slice = tuple(
+        slice(None) if i in area_dim else None
+        for i in range(cube.ndim)
+    )
+
+    # Calculate grid cell volume as area * thickness
+    grid_volume = area.lazy_data()[area_slice] * thickness[z_slice]
```

Comment:
Thanks, it's much simpler. The only question I have is if this iris function keeps the data lazy. Checking the iris source code, it seems that it doesn't, but I haven't checked. 

In any case, maybe it is better to use this function and ask iris to make the changes?

Reply:
Yes, since [Iris 3.7](https://scitools-iris.readthedocs.io/en/latest/whatsnew/3.7.html#features), this function maintains lazy data. If it doesn't do that locally for you, you might need to update your installation.

Reply:
My bad, I didn't check the documentation for the latest stable version! Thanks for that :)